### PR TITLE
Sharing Playlists

### PR DESCRIPTION
**Pull request by akrol95**

### CODE REVIEW

File: BMM.Core/Implementations/DeepLinking/Base/DeepLinkParserBase.cs [hunk 1]
```diff
@@ -0,0 +1,24 @@
+using System;
+using System.Threading.Tasks;
+
+namespace BMM.Core.Implementations.DeepLinking.Base
+{
+    public abstract class DeepLinkParserBase
```
Comment (kkuepper) [inline BMM.Core/Implementations/DeepLinking/Base/DeepLinkParserBase.cs:6]: Hm. Not sure about this one.
I will need to look at the rest and will come back to this.

File: BMM.Core/Implementations/DeepLinking/Base/Interfaces/IDeepLinkParameters.cs [hunk 2]
```diff
@@ -0,0 +1,6 @@
+namespace BMM.Core.Implementations.DeepLinking.Base.Interfaces
+{
+    public interface IDeepLinkParameters
```
Comment (kkuepper) [inline BMM.Core/Implementations/DeepLinking/Base/Interfaces/IDeepLinkParameters.cs:3]: empty interface?

File: BMM.Core/Implementations/DeepLinking/DeepLinkHandler.cs [hunk 3]
```diff
@@ -71,12 +75,13 @@ IUserAuthChecker authChecker
                 new RegexDeepLink("^/music$", NavigateTo<ExploreRecentMusicViewModel>),
                 new RegexDeepLink("^/contributors$", NavigateTo<ExploreContributorsViewModel>),
                 new RegexDeepLink("^/featured$", NavigateTo<CuratedPlaylistsViewModel>),
-                new RegexDeepLinkWithParameters("^/playlist/curated/(?<id>[0-9]+)(/(?<name>.*))?$", OpenCuratedPlaylist),
-                new RegexDeepLinkWithParameters("^/playlist/private/(?<id>[0-9]+)(/(?<name>.*))?$", OpenTrackCollection),
-                new RegexDeepLinkWithParameters("^/playlist/contributor/(?<id>[0-9]+)(/(?<name>.*))?$", OpenContributor),
-                new RegexDeepLinkWithParameters("^/playlist/podcast/(?<id>[0-9]+)(/(?<name>.*))?$", OpenPodcast),
-                new RegexDeepLinkWithParameters("^/album/(?<id>[0-9]+)$", OpenAlbum),
-                new TrackLinkParser("^/track/(?<id>[0-9]+)(/(?<language>.*))?$", (id, name, startTimeInMs) => PlayTrackById(id, startTimeInMs)),
+                new RegexDeepLink<StandardDeepLinkParameters>("^/playlist/curated/(?<id>[0-9]+)(/(?<name>.*))?$", OpenCuratedPlaylist),
```
Comment (kkuepper) [inline BMM.Core/Implementations/DeepLinking/DeepLinkHandler.cs:78]: I think the names "StandardDeepLinkParameters" and "TrackCollectionDeepLinkParameters" can be improved

File: BMM.Core/Implementations/DeepLinking/Base/Interfaces/IDeepLinkParameters.cs (hunk 2 above)
Comment (kkuepper) [reply BMM.Core/Implementations/DeepLinking/Base/Interfaces/IDeepLinkParameters.cs:3]: I think I get it now

File: BMM.UI.Android/Resources/layout/listitem_tracklist_header.xml [hunk 4]
```diff
@@ -43,19 +43,37 @@
             android:layout_marginRight="16dp"
             android:layout_marginBottom="4dp"
             android:gravity="center"
-            android:textFontWeight="800"
```
Comment (kkuepper) [inline BMM.UI.Android/Resources/layout/listitem_tracklist_header.xml:46]: is this not needed?

File: BMM.UI.iOS/Constants/Typography.cs [hunk 5]
```diff
@@ -20,16 +20,24 @@ public static class Typography
             return font;
         });
 
-        public static readonly Lazy<UIFont> Header2 = new Lazy<UIFont>(() =>
+        public static readonly Lazy<UIFont> Header2= new Lazy<UIFont>(() =>
```
Comment (kkuepper) [inline BMM.UI.iOS/Constants/Typography.cs:23]: why did you remove the space?

**Review by kkuepper — CHANGES_REQUESTED** [review]
@olivereisenhut should have a look. Especially the iOS parts.
In general it's good. I just have some comments about details.

File: BMM.UI.iOS/Application/ViewController/EditTrackCollectionViewController.cs [hunk 6]
```diff
@@ -27,7 +27,11 @@ public override void ViewDidLoad()
         {
             base.ViewDidLoad();
 
-            NavigationController.PresentationController.Delegate = new EditTrackCollectionPresentationControllerDelegate(ViewModel);
+            NavigationController.PresentationController.Delegate = new CustomUIAdaptivePresentationControllerDelegate()
+            {
+                OnDidDismiss = HandleDismiss,
+                OnDidAttemptToDismiss = HandleDismiss
```
Comment (olivereisenhut) [inline BMM.UI.iOS/Application/ViewController/EditTrackCollectionViewController.cs:33]: Here it shouldn't invalidate the delegate actions. With the current implementation the dialog is just shown once when trying to dismiss.

**Review by olivereisenhut — CHANGES_REQUESTED** [review]

File: BMM.Core/Implementations/DeepLinking/DeepLinkHandler.cs (hunk 3 above)
Comment (akrol95) [reply BMM.Core/Implementations/DeepLinking/DeepLinkHandler.cs:78]: I aggree with "SharingSecretParameter" but not with "IdAndNameParameter". What if in the future we add another value that is used there? Will we rename it to "IdAndNameAnd(...)Parameter"? ;)
I named it Standard, because it is the most used set of parameters in deep links iin our case.

File: BMM.Core/Implementations/DeepLinking/RegexDeepLink'.cs [hunk 7]
```diff
@@ -0,0 +1,52 @@
+﻿using System;
```
[the remark this answers is not in the document]
Comment (akrol95) [reply BMM.Core/Implementations/DeepLinking/RegexDeepLink'.cs:1]: It it generic class for RegexDeepLink.

Do we have some convention of keeping classes with the same name but with diferent generic parameters? I usually extract it to separate file with the same name and ' at the end.

But if you want, I can keep it in one file.

File: BMM.Core/Implementations/DeepLinking/Parameters/TrackCollectionDeepLinkParameters.cs [hunk 8]
```diff
@@ -0,0 +1,9 @@
+using BMM.Core.Implementations.DeepLinking.Base.Interfaces;
+
+namespace BMM.Core.Implementations.DeepLinking.Parameters
+{
+    public class TrackCollectionDeepLinkParameters : IDeepLinkParameters
```
[the remark this answers is not in the document]
Comment (akrol95) [reply BMM.Core/Implementations/DeepLinking/Parameters/TrackCollectionDeepLinkParameters.cs:5]: I can do it, but I prefer do do it it separate files. I this it's cleaner, especially that I store all the classes for parameters in one folder.

What do you think?

File: BMM.Core/Translation/en/main.json [hunk 9]
```diff
@@ -178,7 +178,11 @@
     "Title": "Favorites",
     "DownloadedContent": "Downloaded Content",
     "FollowedPodcasts": "Followed Podcasts",
-    "MyTracks": "My Tracks"
+    "MyTracks": "My Tracks",
+    "MyPlaylists": "My playlists",
+    "SharedWithMe": "Shared with me",
+    "TrackCountFormat": "{0} Tracks",
```
[the remark this answers is not in the document]
Comment (akrol95) [reply BMM.Core/Translation/en/main.json:184]: Do you mean that i should introduce Plural And Singular form of this format or that I need to use DocumentsViewModel.PluralTracks entry instead of creating new one (there's difference between them in letter size, but  i'm not sure of we care about it in this case)?

File: BMM.Core/Translation/en/main.json [hunk 10]
```diff
@@ -332,6 +338,22 @@
...
   },
+  "ShareTrackCollectionViewModel": {
+    "Title": "Share Playlist",
+    "Done": "Done",
+    "ShareNote": "By sharing a playlist, people will be able to access and listen to the tracks you add, if they have the link.",
+    "ShareLink": "Share link",
+    "MakePrivate": "Make private",
+    "Private": "Private",
+    "SharedWithFormat": "Shared with {0} people",
+    "By": "by"
```
[the remark this answers is not in the document]
Comment (akrol95) [reply BMM.Core/Translation/en/main.json:349]: There's a difference in letters size. I can change it from code but I'd rather not do it because im not sure if this won't brake different languages grammatical rules.

File: BMM.UI.Android/Resources/layout/listitem_tracklist_header.xml (hunk 4 above)
Comment (akrol95) [reply BMM.UI.Android/Resources/layout/listitem_tracklist_header.xml:46]: Because it is in Style right now.

File: BMM.Core/Translation/en/main.json [hunk 11]
```diff
@@ -387,6 +409,8 @@
     "Contributor.Share": "Share contributor",
     "TrackCollection.DeletePlaylist": "Delete playlist",
     "TrackCollection.EditPlaylist": "Edit playlist",
+    "TrackCollection.SharePlaylist": "Share playlist",
```
[the remark this answers is not in the document]
Comment (akrol95) [reply BMM.Core/Translation/en/main.json:412]: Fixed along with the rest of duplicated keys

File: BMM.Api/Implementation/Models/TrackCollection.cs [hunk 12]
```diff
@@ -10,6 +10,10 @@ public class TrackCollection : BaseTrackCollection
 
         public int TrackCount { get => Tracks?.Count > 0 ? Tracks.Count : _trackCount; set => _trackCount = value; }
 
+        public int FollowerCount { get; set; }
+        public string ShareLink { get; set; }
+        public string AuthorName { get; set; }
+        public bool CanEdit { get; set; }
         public List<Track> Tracks { get; set; }
```
[the remark this answers is not in the document]
Comment (akrol95) [reply BMM.Api/Implementation/Models/TrackCollection.cs:17]: You mean spaces between properties?

File: BMM.Core/Models/Enums/DocumentGroupType.cs [hunk 13]
```diff
@@ -0,0 +1,8 @@
+namespace BMM.Core.Models.Enums
+{
+    public enum DocumentGroupType
+    {
+        MyPlaylist,
+        SharedWithMe
+    }
```
[the remark this answers is not in the document]
Comment (akrol95) [reply BMM.Core/Models/Enums/DocumentGroupType.cs:7]: Yes, removed

File: BMM.UI.iOS/Application/Extensions/ViewExtensions.cs [hunk 14]
```diff
@@ -0,0 +1,19 @@
+using UIKit;
+
+namespace BMM.UI.iOS.Extensions
+{
+    public static class ViewExtensions
+    {
+        /// <summary>
+        /// This method should be used for changing Hidden state in views added to UIStackView
+        /// as this property needs to be guarded of multiple set the same state.
```
[the remark this answers is not in the document]
Comment (akrol95) [reply BMM.UI.iOS/Application/Extensions/ViewExtensions.cs:9]: if you set IsHidden two times to False and then set to True, the control will stay visible.
Take a look: http://www.openradar.me/25087688

File: BMM.Core/Implementations/DeepLinking/DeepLinkHandler.cs (hunk 3 above)
Comment (kkuepper) [reply BMM.Core/Implementations/DeepLinking/DeepLinkHandler.cs:78]: But why would we add another parameter? It seems very unlikely that we add a parameter to all playlists at once. If we need another parameter for podcast, then we would need another IDeepLinkParameter class and then we can pick the name that makes sense. But the existing class IdAndNameParameter would stay the same.

File: BMM.Core/Implementations/DeepLinking/Parameters/TrackCollectionDeepLinkParameters.cs (hunk 8 above)
[the remark this answers is not in the document]
Comment (kkuepper) [reply BMM.Core/Implementations/DeepLinking/Parameters/TrackCollectionDeepLinkParameters.cs:5]: sounds good. especially since they are in 1 folder

File: BMM.Core/Translation/en/main.json (hunk 9 above)
[the remark this answers is not in the document]
Comment (kkuepper) [reply BMM.Core/Translation/en/main.json:184]: I think it should be the same entry.
Also if you look at the design it's always using lowercase.

File: BMM.Core/ViewModels/AlbumViewModel.cs [hunk 15]
```diff
@@ -133,6 +133,8 @@ private void UpdateView(object sender, NotifyCollectionChangedEventArgs e)
 
         public bool ShowTrackCount => true;
 
+        public bool ShowAddToPlaylistButton => false;
```
[the remark this answers is not in the document]
Comment (akrol95) [reply BMM.Core/ViewModels/AlbumViewModel.cs:136]: It's only about "add to playlist" butotn on shared playlist screen which is using fragment_tracklist (TBH i'm not a fan of this approach ;) ). I'll change the name to not introduce confusion.

[the remark this answers is not in the document]
Comment (kkuepper) [reply BMM.Core/ViewModels/AlbumViewModel.cs:136]: I remember that the name confused me as well. Good choice to rename it

File: BMM.UI.iOS/Application/ViewController/EditTrackCollectionViewController.cs (hunk 6 above)
Comment (akrol95) [reply BMM.UI.iOS/Application/ViewController/EditTrackCollectionViewController.cs:33]: Sorry, I don't get it. Could you elaborate more?

Comment (olivereisenhut) [reply BMM.UI.iOS/Application/ViewController/EditTrackCollectionViewController.cs:33]: In the edit view after a change the first time you try to drag the sheet down it shakes and asks you if you want to keep editing. With this change it's just happening once. When I try to drag the sheet down again it's not asking if you want keep editing or discard the changes. It should ask you every time if you have changes if you want to keep editing or discard.

File: BMM.Api/Implementation/Models/TrackCollection.cs (hunk 12 above)
[the remark this answers is not in the document]
Comment (olivereisenhut) [reply BMM.Api/Implementation/Models/TrackCollection.cs:17]: Yep. Actually I thought I might be configured in the reshaper settings but I might be wrong.

File: BMM.UI.iOS/Application/ViewController/EditTrackCollectionViewController.cs (hunk 6 above)
Comment (akrol95) [reply BMM.UI.iOS/Application/ViewController/EditTrackCollectionViewController.cs:33]: Ok, now I see. I haven't noticed "not" word in your github comment ;) https://github.com/bcc-code/bmm-app/pull/37#discussion_r661233186

File: BMM.Core/ViewModels/Parameters/Interface/ITrackCollectionParameter.cs [hunk 16]
```diff
@@ -0,0 +1,9 @@
+namespace BMM.Core.ViewModels.Parameters.Interface
+{
+    public interface ITrackCollectionParameter
+    {
+        int? TrackCollectionId { get; }
+        string Name { get; }
+        string SharingSecret { get; }
```
Comment (kkuepper) [inline BMM.Core/ViewModels/Parameters/Interface/ITrackCollectionParameter.cs:7]: I think it's better to have the SharingSecret separately, e.g. in a SharingSecretParameter

File: BMM.Core/ViewModels/Parameters/TrackCollectionParameter.cs [hunk 17]
```diff
@@ -0,0 +1,21 @@
+using BMM.Core.ViewModels.Parameters.Interface;
+
+namespace BMM.Core.ViewModels.Parameters
+{
+    public class TrackCollectionParameter : ITrackCollectionParameter
+    {
+        public TrackCollectionParameter(
+            int? trackCollectionId = null,
```
Comment (kkuepper) [inline BMM.Core/ViewModels/Parameters/TrackCollectionParameter.cs:8]: by moving the SharingSecret into it's own Parameter the trackCollectionId doesn't need to be nullable.

File: BMM.UI.iOS/Application/ViewController/EditTrackCollectionViewController.cs (hunk 6 above)
Comment (kkuepper) [reply BMM.UI.iOS/Application/ViewController/EditTrackCollectionViewController.cs]: you never addressed Oliver's comment

**Review by kkuepper — APPROVED** [review]
